core/src/main/scala/org/apache/spark/sql/execution/columnar/CachedBatchCreator.scala

```diff
@@ -16,150 +16,32 @@
-        // type, you get through a BigDecimal object with scale as 0. But the
-        // dataframe schema has correct type as DecimalType(12, 2).
-        // Thus, after saving the dataframe into parquet file and then
-        // retrieve it, you will get wrong result 199.99. So it is needed
-        // to set precision and scale for Decimal based on JDBC metadata.
-        case d: DecimalType =>
-          val dvd = execRow.getColumn(pos)
-          if (dvd == null || dvd.isNull) {
-            mutableRow.update(i, null)
-          } else {
-            val p = d.precision
-            val s = d.scale
-            dvd.typeToBigDecimal() match {
-              case Types.DECIMAL => mutableRow.update(i,
-                Decimal(dvd.getObject.asInstanceOf[java.math.BigDecimal], p, s))
-              case Types.CHAR => mutableRow.update(i,
-                Decimal(BigDecimal(dvd.getString), p, s))
-              case Types.BIGINT => mutableRow.update(i,
-                Decimal(dvd.getLong, p, s))
-              case o => throw new IllegalArgumentException(
-                s"Unsupported typeToBigDecimal result = $o")
-            }
-          }
-        case DoubleType =>
-          mutableRow.setDouble(i, execRow.getColumn(pos).getDouble)
-        case FloatType =>
-          mutableRow.setFloat(i, execRow.getColumn(pos).getFloat)
-        case TimestampType =>
-          val t = execRow.getColumn(pos).getTimestamp(null)
-          if (t != null) {
-            mutableRow.setLong(i, DateTimeUtils.fromJavaTimestamp(t))
-          } else {
-            mutableRow.update(i, null)
-          }
-        case BinaryType =>
-          mutableRow.update(i, execRow.getColumn(pos).getBytes)
-        case _: ArrayType =>
-          val bytes = execRow.getColumn(pos).getBytes
-          val array = new UnsafeArrayData
-          array.pointTo(bytes, Platform.BYTE_ARRAY_OFFSET, bytes.length)
-          mutableRow.update(i, array)
-        case _: MapType =>
-          val bytes = execRow.getColumn(pos).getBytes
-          val map = new UnsafeMapData
-          map.pointTo(bytes, Platform.BYTE_ARRAY_OFFSET, bytes.length)
-          mutableRow.update(i, map)
-        case s: StructType =>
-          val bytes = execRow.getColumn(pos).getBytes
-          val row = new UnsafeRow
-          row.pointTo(bytes, Platform.BYTE_ARRAY_OFFSET,
-            s.fields.length, bytes.length)
-          mutableRow.update(i, row)
-        case _ => throw new IllegalArgumentException(
-          s"Unsupported field ${schema.fields(i)}")
-      }
-      if (execRow.getColumn(pos).isNull) {
-        mutableRow.setNullAt(i)
-      }
-      i = i + 1
-    }
-    mutableRow
-  }
+    val useCompression: Boolean) extends CompactExecRowToMutableRow {
 
   def createAndStoreBatch(sc: ScanController, row: AbstractCompactExecRow,
-      batchID: UUID, bucketID: Int): mutable.HashSet[Any] = {
+      batchID: UUID, bucketID: Int): mutable.HashSet[AnyRef] = {
```

Comment:
This return type looks a bit redundant. As we are converting them again to Java types by JavaConversions, we can get the Java types at the first place. Can reduce one conversion.